modules/core/src/test/java/org/apache/ignite/internal/processors/cache/eviction/EvictionPolicyFailureHandlerTest.java

```diff
@@ -0,0 +1,223 @@
+
+        IgniteEx node = startGrid(0);
+
+        IgniteEx client = startClientGrid(1);
+
+        GridCacheAdapter<Object, Object> cache = ((IgniteKernal)node).internalCache(DEFAULT_CACHE_NAME);
+
+        cache.put(1, 1);
+
+        CountDownLatch locPeekFinished = new CountDownLatch(1);
+        CountDownLatch txStarted = new CountDownLatch(1);
+        CountDownLatch txFinished = new CountDownLatch(1);
+
+        GridTestUtils.runAsync(() -> {
+            IgniteCache<Object, Object> cache1 = client.cache(DEFAULT_CACHE_NAME);
+
+            IgniteTransactions transactions = client.transactions();
+
+            try (Transaction tx = transactions.txStart(PESSIMISTIC, REPEATABLE_READ)) {
+                cache1.put(2.1, 2.4);
+
+                txStarted.countDown();
+
+                locPeekFinished.await();
+
+                tx.commit();
+            }
+            catch (Exception ignore) {
+            }
+
+            txFinished.countDown();
+
+        }, "tx-thread");
+
+        txStarted.await();
+
+        try {
+            cache.localPeek(2.1, new CachePeekMode[] {CachePeekMode.ONHEAP});
+        }
+        catch (Exception ignore) {
+        }
+
+        locPeekFinished.countDown();
+
+        assertTrue(lsnr.check(10_000));
+
+        txFinished.await();
+
+        assertFalse(cache.map().entrySet(cache.context().cacheId()).stream()
+            .anyMatch(e -> new Double(2.1).equals(e.key().value(null, false)))
+        );
+
+        assertTrue(node.cluster().active());
+    }
+
+    /**
+     * We skip {@link RuntimeException}, but {@link Error} should trigger {@link FailureHandler}.
+     */
+    @Test
+    public void testErrorShouldCallErrorHandler() throws Exception {
+        oom = true;
+
+        IgniteEx node1 = startGrid(0);
+
+        IgniteEx node2 = startGrid(1);
+
+        IgniteCache<Object, Object> cache = node2.cache(DEFAULT_CACHE_NAME);
+
+        Affinity<Object> affinity = affinity(cache);
+
+        try {
+            for (int i = 0; i < 1000; i++) {
+                if (affinity.isPrimary(node1.localNode(), i))
+                    cache.put(i, 1);
+            }
+        }
+        catch (Throwable ignore) {
+        }
+
+        assertTrue(nodeFailure.get());
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void testFailureHandlerShouldNotCallOnRuntimeException() throws Exception {
+        IgniteEx node1 = startGrid(0);
+
+        IgniteEx node2 = startGrid(1);
+
+        IgniteCache<Object, Object> cache = node2.cache(DEFAULT_CACHE_NAME);
+
+        Affinity<Object> affinity = affinity(cache);
+
+        for (int i = 0; i < 1000; i++) {
+            if (affinity.isPrimary(node1.localNode(), i))
+                cache.put(i, 1);
+        }
+
+        for (int i = 0; i < 1000; i++) {
+            double d = (double)i;
+
+            if (affinity.isPrimary(node1.localNode(), d))
+                cache.put(d, 1);
+        }
+
+        assertFalse(nodeFailure.get());
```

Comment:
Could you please add extra check in order to validate that all "double-typed" entries were evicted from on-heap?

Reply:
Yes, I added it.